// roamjs-components/util/extensionDeprecatedWarning.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const idToTitle_1 = tslib_1.__importDefault(require("./idToTitle"));
const SimpleAlert_1 = require("../components/SimpleAlert");
const getRoamUrl_1 = tslib_1.__importDefault(require("../dom/getRoamUrl"));
const deleteBlock_1 = tslib_1.__importDefault(require("../writes/deleteBlock"));
const Toast_1 = require("../components/Toast");
const getBasicTreeByParentUid_1 = tslib_1.__importDefault(require("../queries/getBasicTreeByParentUid"));
const getPageUidByPageTitle_1 = tslib_1.__importDefault(require("../queries/getPageUidByPageTitle"));
const getSubTree_1 = tslib_1.__importDefault(require("./getSubTree"));
const extensionDeprecatedWarning = ({ extensionId, reason, }) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const configUid = (0, getPageUidByPageTitle_1.default)(extensionId);
    const config = (0, getBasicTreeByParentUid_1.default)(configUid);
    const donotShowAgainUid = (0, getSubTree_1.default)({
        tree: config,
        key: "Do not show again",
    }).uid;
    if (!donotShowAgainUid) {
        const blocks = window.roamAlphaAPI.data.fast.q(`[:find (pull ?roamjs [:block/uid]) :where [?block :block/string ?contents] [(clojure.string/includes? ?contents  "https://roamjs.com/${extensionId}")] [?roamjs :block/children ?block]]`).map(([block]) => block[":block/uid"] || "");
        (0, SimpleAlert_1.render)({
            content: `RoamJS will soon be deprecating and then removing the ${(0, idToTitle_1.default)(extensionId)} extension.

${reason}

If you no longer use this extension, feel free to uninstall it${blocks.length
                ? ` by removing the block from [here](${(0, getRoamUrl_1.default)(blocks[0])}) or by clicking confirm`
                : ""}. If this will be an issue for any reason, please reach out to support@roamjs.com.`,
            onConfirm: () => blocks.map((b) => (0, deleteBlock_1.default)(b)),
            onCancel: () => (0, Toast_1.render)({
                content: "Reach out to supprt@roamjs.com if you would like to keep the extension.",
                id: `deprecated-${extensionId}`,
            }),
            confirmText: "Confirm",
            dontShowAgain: configUid,
        });
    }
});
exports.default = extensionDeprecatedWarning;
