// @samepage/testing/playwright.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const dotenv_1 = tslib_1.__importDefault(require("dotenv"));
dotenv_1.default.config();
const config = {
    globalSetup: require.resolve("./globalSetup"),
    globalTeardown: require.resolve("./globalTeardown"),
    reporter: process.env.CI
        ? [["github"], ["html", { outputFolder: "playwright-report" }]]
        : [["html", { open: "on-failure" }]],
    testDir: `${process.cwd()}/tests`,
    use: {
        screenshot: "only-on-failure",
        video: process.env.DEBUG || process.env.VIDEO
            ? "on"
            : process.env.CI
                ? "retry-with-video"
                : "off",
    },
    projects: [
        // integration - your code is run as a separate process for black box testing
        { name: "integration", testMatch: /integration/ },
        // unit - your code is run in the same process for granular logic testing
        { name: "unit", testMatch: /^((?!integration).)*$/ },
    ],
};
exports.default = config;
