// aws-sdk-plus/dist/errors.js
"use strict";
// https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses
Object.defineProperty(exports, "__esModule", { value: true });
exports.InternalServorError = exports.UnavailableForLegalReasonsError = exports.RequestHeaderFieldsTooLargeError = exports.TooManyRequestsError = exports.PreconditionRequiredError = exports.UpgradeRequiredError = exports.TooEarlyError = exports.FailedDependencyError = exports.LockedError = exports.UnprocessableEntityError = exports.MisdirectedRequestError = exports.IAmATeapotError = exports.ExpectationFailedError = exports.RangeNotSatisfiableError = exports.UnsupportedMediaTypeError = exports.URITooLongError = exports.PayloadTooLargeError = exports.PreconditionFailedError = exports.LengthRequiredError = exports.GoneError = exports.ConflictError = exports.RequestTimeoutError = exports.ProxyAuthenticationRequiredError = exports.NotAcceptableError = exports.MethodNotAllowedError = exports.NotFoundError = exports.ForbiddenError = exports.PaymentRequiredError = exports.UnauthorizedError = exports.BadRequestError = void 0;
/**
 * The server could not understand the request due to invalid syntax.
 */
class BadRequestError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 400;
    }
}
exports.BadRequestError = BadRequestError;
/**
 * Although the HTTP standard specifies "unauthorized", semantically this response means "unauthenticated".
 * That is, the client must authenticate itself to get the requested response.
 */
class UnauthorizedError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 401;
    }
}
exports.UnauthorizedError = UnauthorizedError;
/**
 * This response code is reserved for future use. The initial aim for creating this code was using
 * it for digital payment systems, however this status code is used very rarely and no standard convention exists.
 */
class PaymentRequiredError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 402;
    }
}
exports.PaymentRequiredError = PaymentRequiredError;
/**
 * The client does not have access rights to the content; that is, it is unauthorized,
 * so the server is refusing to give the requested resource. Unlike 401 Unauthorized, the client's
 * identity is known to the server.
 */
class ForbiddenError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 403;
    }
}
exports.ForbiddenError = ForbiddenError;
/**
 * The server can not find the requested resource. In the browser, this means the URL is not recognized.
 * In an API, this can also mean that the endpoint is valid but the resource itself does not exist.
 * Servers may also send this response instead of 403 Forbidden to hide the existence of a
 * resource from an unauthorized client. This response code is probably the most well known due to
 * its frequent occurrence on the web.
 */
class NotFoundError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 404;
    }
}
exports.NotFoundError = NotFoundError;
/**
 * The request method is known by the server but is not supported by the target resource.
 * For example, an API may not allow calling DELETE to remove a resource.
 */
class MethodNotAllowedError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 405;
    }
}
exports.MethodNotAllowedError = MethodNotAllowedError;
/**
 *  This response is sent when the web server, after performing server-driven content negotiation, doesn't find any content that conforms to the criteria given by the user agent.
 */
class NotAcceptableError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 406;
    }
}
exports.NotAcceptableError = NotAcceptableError;
/**
 *  This is similar to 401 Unauthorized but authentication is needed to be done by a proxy.
 */
class ProxyAuthenticationRequiredError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 407;
    }
}
exports.ProxyAuthenticationRequiredError = ProxyAuthenticationRequiredError;
/**
 *  This response is sent on an idle connection by some servers, even without any previous request by the client. It means that the server would like to shut down this unused connection. This response is used much more since some browsers, like Chrome, Firefox 27+, or IE9, use HTTP pre-connection mechanisms to speed up surfing. Also note that some servers merely shut down the connection without sending this message.
 */
class RequestTimeoutError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 408;
    }
}
exports.RequestTimeoutError = RequestTimeoutError;
/**
 *  This response is sent when a request conflicts with the current state of the server.
 */
class ConflictError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 409;
    }
}
exports.ConflictError = ConflictError;
/**
 *  This response is sent when the requested content has been permanently deleted from server, with no forwarding address. Clients are expected to remove their caches and links to the resource. The HTTP specification intends this status code to be used for "limited-time, promotional services". APIs should not feel compelled to indicate resources that have been deleted with this status code.
 */
class GoneError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 410;
    }
}
exports.GoneError = GoneError;
/**
 *  Server rejected the request because the Content-Length header field is not defined and the server requires it.
 */
class LengthRequiredError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 411;
    }
}
exports.LengthRequiredError = LengthRequiredError;
/**
 *  The client has indicated preconditions in its headers which the server does not meet.
 */
class PreconditionFailedError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 412;
    }
}
exports.PreconditionFailedError = PreconditionFailedError;
/**
 *  Request entity is larger than limits defined by server. The server might close the connection or return an Retry-After header field.
 */
class PayloadTooLargeError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 413;
    }
}
exports.PayloadTooLargeError = PayloadTooLargeError;
/**
 *  The URI requested by the client is longer than the server is willing to interpret.
 */
class URITooLongError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 414;
    }
}
exports.URITooLongError = URITooLongError;
/**
 *  The media format of the requested data is not supported by the server, so the server is rejecting the request.
 */
class UnsupportedMediaTypeError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 415;
    }
}
exports.UnsupportedMediaTypeError = UnsupportedMediaTypeError;
/**
 *  The range specified by the Range header field in the request cannot be fulfilled. It's possible that the range is outside the size of the target URI's data.
 */
class RangeNotSatisfiableError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 416;
    }
}
exports.RangeNotSatisfiableError = RangeNotSatisfiableError;
/**
 *  This response code means the expectation indicated by the Expect request header field cannot be met by the server.
 */
class ExpectationFailedError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 417;
    }
}
exports.ExpectationFailedError = ExpectationFailedError;
/**
 *  The server refuses the attempt to brew coffee with a teapot.
 */
class IAmATeapotError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 418;
    }
}
exports.IAmATeapotError = IAmATeapotError;
/**
 *  The request was directed at a server that is not able to produce a response. This can be sent by a server that is not configured to produce responses for the combination of scheme and authority that are included in the request URI.
 */
class MisdirectedRequestError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 421;
    }
}
exports.MisdirectedRequestError = MisdirectedRequestError;
/**
 *  The request was well-formed but was unable to be followed due to semantic errors.
 */
class UnprocessableEntityError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 422;
    }
}
exports.UnprocessableEntityError = UnprocessableEntityError;
/**
 *  The resource that is being accessed is locked.
 */
class LockedError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 423;
    }
}
exports.LockedError = LockedError;
/**
 *  The request failed due to failure of a previous request.
 */
class FailedDependencyError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 424;
    }
}
exports.FailedDependencyError = FailedDependencyError;
/**
 *  Indicates that the server is unwilling to risk processing a request that might be replayed.
 */
class TooEarlyError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 425;
    }
}
exports.TooEarlyError = TooEarlyError;
/**
 *  The server refuses to perform the request using the current protocol but might be willing to do so after the client upgrades to a different protocol. The server sends an Upgrade header in a 426 response to indicate the required protocol(s).
 */
class UpgradeRequiredError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 426;
    }
}
exports.UpgradeRequiredError = UpgradeRequiredError;
/**
 *  The origin server requires the request to be conditional. This response is intended to prevent the 'lost update' problem, where a client GETs a resource's state, modifies it and PUTs it back to the server, when meanwhile a third party has modified the state on the server, leading to a conflict.
 */
class PreconditionRequiredError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 428;
    }
}
exports.PreconditionRequiredError = PreconditionRequiredError;
/**
 *  The user has sent too many requests in a given amount of time ("rate limiting").
 */
class TooManyRequestsError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 429;
    }
}
exports.TooManyRequestsError = TooManyRequestsError;
/**
 *  The server is unwilling to process the request because its header fields are too large. The request may be resubmitted after reducing the size of the request header fields.
 */
class RequestHeaderFieldsTooLargeError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 431;
    }
}
exports.RequestHeaderFieldsTooLargeError = RequestHeaderFieldsTooLargeError;
/**
 *  The user agent requested a resource that cannot legally be provided, such as a web page censored by a government.
 */
class UnavailableForLegalReasonsError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 451;
    }
}
exports.UnavailableForLegalReasonsError = UnavailableForLegalReasonsError;
/**
 * The server has encountered a situation it does not know how to handle.
 */
class InternalServorError extends Error {
    constructor(arg) {
        super(arg);
        this.code = 500;
    }
}
exports.InternalServorError = InternalServorError;
