// aws-sdk-plus/dist/sendEmail.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const aws_sdk_1 = __importDefault(require("aws-sdk"));
const server_1 = __importDefault(require("react-dom/server"));
const ses = new aws_sdk_1.default.SES();
const support = process.env.SUPPORT_EMAIL ||
    (process.env.HOST
        ? `support@${process.env.HOST.replace(/^https?:\/\//, "")}`
        : "");
const sendEmail = ({ to = support, body, subject, from = support, replyTo, }) => ses
    .sendEmail({
    Destination: {
        ToAddresses: typeof to === "string" ? [to] : to,
    },
    Message: {
        Body: {
            Html: {
                Charset: "UTF-8",
                Data: typeof body === "string"
                    ? body
                    : server_1.default.renderToStaticMarkup(body),
            },
        },
        Subject: {
            Charset: "UTF-8",
            Data: subject,
        },
    },
    Source: from,
    ReplyToAddresses: typeof replyTo === "string" ? [replyTo] : replyTo,
})
    .promise()
    .then((r) => r.MessageId);
exports.default = sendEmail;
