// roamjs-components/components/ProgressDialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.render = exports.ID = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@blueprintjs/core");
const react_1 = tslib_1.__importStar(require("react"));
const createOverlayRender_1 = tslib_1.__importDefault(require("../util/createOverlayRender"));
exports.ID = "roamjs-progress-dialog-root";
const ProgressDialog = (_a) => {
    var { onClose } = _a, props = tslib_1.__rest(_a, ["onClose"]);
    const [maxTimeout, setMaxtimeout] = (0, react_1.useState)(props.timeout);
    const [now, setNow] = (0, react_1.useState)(new Date());
    const onId = (0, react_1.useCallback)((props) => {
        window.roamjs.actions[props.id] = props.timeout;
        setMaxtimeout(props.timeout);
    }, [setMaxtimeout]);
    (0, react_1.useEffect)(() => {
        const element = document.getElementById(exports.ID);
        if (element) {
            element.addEventListener("log", ((e) => {
                onId(e.detail);
            }));
            onId(props);
        }
        const interval = setInterval(() => setNow(new Date()), 1000);
        return () => clearInterval(interval);
    }, [onId, setTimeout, props]);
    const numActions = Object.keys(window.roamjs.actions).length;
    (0, react_1.useEffect)(() => {
        if (numActions === 0)
            onClose();
    }, [numActions]);
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement("style", null, `.roamjs-progress-dialog {
  right: 8px;
  margin: 0;
  position: absolute;
  bottom: 8px;
}
.roamjs-progress-dialog-portal .bp3-dialog-container,
.roamjs-progress-dialog-portal .bp3-overlay {
  pointer-events: none;
}
.roamjs-progress-dialog-portal .bp3-overlay-backdrop {
  display: none;
}`),
        react_1.default.createElement(core_1.Dialog, { isOpen: true, onClose: onClose, portalClassName: "roamjs-progress-dialog-portal", className: "roamjs-progress-dialog", canOutsideClickClose: false, canEscapeKeyClose: false, hasBackdrop: false, enforceFocus: false },
            react_1.default.createElement("div", { style: { padding: 16 } },
                react_1.default.createElement("h4", null, "Performing Writes to Roam..."),
                react_1.default.createElement("p", null,
                    "Still have ",
                    react_1.default.createElement("b", null, Object.keys(window.roamjs.actions).length),
                    " ",
                    "actions to write. Expected to finish the last one in",
                    " ",
                    Math.ceil((maxTimeout - now.valueOf()) / 1000),
                    " seconds...")))));
};
exports.render = (0, createOverlayRender_1.default)("progress-dialog", ProgressDialog);
exports.default = ProgressDialog;
